### PR TITLE
[package-extractor]Add dependencies configurations field in package extractor

### PR DESCRIPTION
**Pull request by EscapeB**
## Summary
Add "dependenciesConfigurations" field in options. 
Related discussion in https://github.com/microsoft/rushstack/pull/4210

## Details
1. Provide "dependenciesConfigurations" field in options to configure behavior when extracting third party dependencies. 
2. Add similar fields named "patternToInclude" and "patternToExclude" to filter files inside third party dependencies.

## How it was tested
Added some unit test cases to test the new field.

### CODE REVIEW

**Comment by EscapeB** [comment]
Hi @D4N14L, could you please help to review this PR?

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (iclanton) [inline libraries/package-extractor/src/PackageExtractor.ts:81]:
```suggestion
  /**
   * As there might be different versions of same dependency, we need take version field into consideration
   * The key format `${name}@${version}`
   */
```

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (iclanton) [inline libraries/package-extractor/src/PackageExtractor.ts:146]: Are these paths? Maybe update the name to something like `filePathPatternsToExclude` and `filePathPatternsToInclude` to be completely clear.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (iclanton) [inline libraries/package-extractor/src/PackageExtractor.ts:701]:
```suggestion
      isFileExcluded(filePath, sourceProjectConfiguration);
```

The second parameter doesn't appear to get modified, so it doesn't need to be cloned.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (iclanton) [inline libraries/package-extractor/src/PackageExtractor.ts:675]:
```suggestion
      { patternsToInclude, patternsToExclude}: Pick<IExtractorDependencyConfiguration, 'patternsToInclude' | 'patternsToExclude'>
```

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (iclanton) [inline libraries/package-extractor/src/PackageExtractor.ts:680]:
```suggestion
      if (patternsToInclude?.length) {
        includeFilters = patternsToInclude?.map((p) => new Minimatch(p, { dot: true }));
```

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (iclanton) [inline libraries/package-extractor/src/PackageExtractor.ts:683]:
```suggestion
      if (patternsToExclude?.length) {
        excludeFilters = patternsToExclude?.map((p) => new Minimatch(p, { dot: true }));
```

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (iclanton) [inline libraries/package-extractor/src/PackageExtractor.ts:715]:
```suggestion
      return isFileExcluded(filePath, dependenciesConfiguration);
```

File: libraries/package-extractor/src/tests/PackageExtractor.test.ts
Comment (iclanton) [inline libraries/package-extractor/src/tests/PackageExtractor.test.ts:1]:
```suggestion
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import path from 'path';
```

File: libraries/package-extractor/src/tests/PackageExtractor.test.ts
Comment (iclanton) [inline libraries/package-extractor/src/tests/PackageExtractor.test.ts:29]: Intentionally commented out?

File: libraries/package-extractor/src/tests/PackageExtractor.test.ts
Comment (iclanton) [inline libraries/package-extractor/src/tests/PackageExtractor.test.ts:14]:
```suggestion
  const testRepoRoot = path.resolve(__dirname, '../../src/tests/package-extractor-test-repo');
```

File: libraries/package-extractor/src/tests/Utils.ts
Comment (iclanton) [inline libraries/package-extractor/src/tests/Utils.ts:1]:
```suggestion
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import path from 'path';
```

File: libraries/package-extractor/src/tests/PackageExtractor.test.ts
Comment (iclanton) [inline libraries/package-extractor/src/tests/PackageExtractor.test.ts:16]: This should be passed in via an argument instead of parsing it again here.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (EscapeB) [reply libraries/package-extractor/src/PackageExtractor.ts:146]: They are glob patterns, same as option filed in projectConfigurations.

File: libraries/package-extractor/src/tests/PackageExtractor.test.ts
Comment (EscapeB) [reply libraries/package-extractor/src/tests/PackageExtractor.test.ts:32]: No, this line should be kept to clean test files generated in unit tests, will fix it

File: libraries/package-extractor/src/tests/PackageExtractor.test.ts
Comment (EscapeB) [reply libraries/package-extractor/src/tests/PackageExtractor.test.ts:19]: I'm not sure how can it be passed in...As it's inside jest test function.
Could you please provide more details?

File: libraries/package-extractor/src/tests/package-extractor-test-repo/packages/foo/node_modules/.pnpm/baz/package.json
Comment (D4N14L) [inline libraries/package-extractor/src/tests/package-extractor-test-repo/packages/foo/node_modules/.pnpm/baz/package.json:1]: The node_modules directory should not be checked in and should be gitignored. If this is required, an install should be performed as a part of the build process.

Comment (D4N14L) [reply libraries/package-extractor/src/tests/package-extractor-test-repo/packages/foo/node_modules/.pnpm/baz/package.json:1]: Ahh hmm, this is intentional. Seems a bit janky of a way to test this... it's not actually an install, so it doesn't have symlinks etc. that would normally exist... Yeah, I would suggest doing an actual install to get the state on-disk to match what you plan to test. Really, you could make two test projects within the repo, and one of them would run the deploy and validate output content. You can add the deploy scenario to the config in the Rushstack repo, so that everything should be prepared for you.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:669]: It feels like there's a lot more going on here than is needed... I have to run, but I'll try to take a closer look tonight to see if there's a way to handle this a bit more simply

File: libraries/package-extractor/src/tests/package-extractor-test-repo/packages/foo/node_modules/.pnpm/baz/package.json
Comment (EscapeB) [reply libraries/package-extractor/src/tests/package-extractor-test-repo/packages/foo/node_modules/.pnpm/baz/package.json:1]: Yes, just created a fake node_modules structure for testing. It had symlinks like pnpm created but without hardlinks to pnpm global store.

1. packages/foo/node_modules/bar - > packages/bar                                                 (local project dependencies)
2. packages/foo/node_modules/baz -> packages/foo/node_modules/.pnpm/baz    (third party dependencies)

Comment (EscapeB) [reply libraries/package-extractor/src/tests/package-extractor-test-repo/packages/foo/node_modules/.pnpm/baz/package.json:1]: Two test projects inside the repo only mocks the local project dependency scenario...

**Comment by D4N14L** [comment]
I just merged my PR which should address your concerns: https://github.com/microsoft/rushstack/pull/4259

I'm a bit iffy on adding general 3rd party pattern include/exclude support given that it's a bit ambiguous if it applies only to the version of the package consumed directly by a project, or another dependency, or otherwise. The original issue should be addressed by my PR though.

**Comment by EscapeB** [comment]
> I just merged my PR which should address your concerns: #4259
> 
> I'm a bit iffy on adding general 3rd party pattern include/exclude support given that it's a bit ambiguous if it applies only to the version of the package consumed directly by a project, or another dependency, or otherwise. The original issue should be addressed by my PR though.

I will rebase my PR branch~

But seems this PR can't solve my original issue, the symlink(which linked to global path) inside third party dependencies still will cause errors :(

**Comment by EscapeB** [comment]
Hi @D4N14L , what I need to do if I want to get this PR merged?

**Comment by D4N14L** [comment]
@EscapeB I'm so sorry that I've let this slide. I've been pretty busy of late and just haven't had a chance to take a closer look. Plan on doing so tomorrow when I can find a moment. Again, apologies for holding this up.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:669]: Use the `sourceFolderRealPath` instead. Looks like that is what's used when it's originally added to the map

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:763]: Move this before the `path.resolve`. It's more expensive than it has any right to be, so we should avoid doing it if we can.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:825]: I think we should drop this, it'll get pretty verbose.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:821]: Both of these depend on the relative source path, Just put them under the same if block

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:142]: We should make this a semver version range

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:321]: If we turn the version into a version range, this will likely have to turn into a `Map<string, IExtractorDependencyConfiguration[]>` which is name -> config, and then cycle through the included configs to check the ranges with semver.

File: common/reviews/api/package-extractor.api.md
Comment (D4N14L) [inline common/reviews/api/package-extractor.api.md:29]:
```suggestion
    dependencyConfigurations?: IExtractorDependencyConfiguration[];
```

**Review by D4N14L — CHANGES_REQUESTED** [review]

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:722]: Combine this with the outer if statement

**Review by D4N14L — CHANGES_REQUESTED** [review]

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [reply libraries/package-extractor/src/PackageExtractor.ts:722]: Better yet, just set the property to `sourceProjectConfiguration && !includeNpmIgnoreFiles`

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (EscapeB) [reply libraries/package-extractor/src/PackageExtractor.ts:142]: Nice idea, have changed to use semver version range.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (EscapeB) [reply libraries/package-extractor/src/PackageExtractor.ts]: LGTM.
But as the Map structure has changed, it can't be easily convert from Array. So I add several lines to covert origin Array to Map.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:743]: You can combine these into the `isFileExcluded` check by checking `isLocalProject` there instead of twice below.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:701]: Since we already know `isLocalProject` and have the package.json, we can just get the correct patterns within this function without needing to pass them in.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [reply libraries/package-extractor/src/PackageExtractor.ts:143]: Let's also rename it to `dependencyVersionRange` as well to make it clear that it accepts a semver range.

File: build-tests/package-extractor-test-03/package.json
Comment (D4N14L) [inline build-tests/package-extractor-test-03/package.json:10]: Curious... I've never seen this version specifier format. How does this work? Why is this needed?

**Review by D4N14L — CHANGES_REQUESTED** [review]
Looking good! Just a few more changes then I think it's good to merge.

File: build-tests/package-extractor-test-03/package.json
Comment (EscapeB) [reply build-tests/package-extractor-test-03/package.json:10]: I saw them in `common/config/rush/common-versions.json`, and the comment says these are versions are used by testing projects.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (EscapeB) [reply libraries/package-extractor/src/PackageExtractor.ts]: Done

**Comment by EscapeB** [comment]
> Looking good! Just a few more changes then I think it's good to merge.

Hi @D4N14L, I think all the changes mentioned in the comments have now been made, please help merge this PR when you have time.

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:84]:
```suggestion
```

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:709]:
```suggestion
        filePath: string,
        patternsToInclude: string[] | undefined,
        patternsToExclude: string[] | undefined
      ): boolean => {
        let includeFilters: IMinimatch[] | undefined;
        let excludeFilters: IMinimatch[] | undefined;
```

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:733]:
```suggestion
        return excludeFileByPatterns(
          filePath,
          sourceProjectConfiguration?.patternsToInclude
          sourceProjectConfiguration?.patternsToExclude
        );
```

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:752]:
```suggestion
        return matchedDependenciesConfigurations.some(
          (d) => excludeFileByPatterns(filePath, d.patternsToInclude, d.patternsToExclude)
        );
```

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:845]:
```suggestion
          if (
            relativeSourcePath !== '' &&
            (ignoreFilter.ignores(relativeSourcePath) || isFileExcluded(relativeSourcePath))
          ) {
            callback();
            return;
          }
```

**Review by D4N14L — APPROVED** [review]

File: libraries/package-extractor/src/PackageExtractor.ts
Comment (D4N14L) [inline libraries/package-extractor/src/PackageExtractor.ts:728]:
```suggestion
          sourceProjectConfiguration?.patternsToInclude,
```

**Pull request merged**
